File: julia/Day11.jl
module Day11

export Submarine, on_input!, output

mutable struct Submarine
    energy_levels::Matrix{UInt32}
    flashed::Matrix{Bool}
    total_flashes::UInt32

    Submarine() = new(Matrix(undef, 0, 0), Matrix(undef, 0, 0), 0)
end

function on_input!(s::Submarine, a::AbstractString)
    energies = parse.(UInt32, split(a, ""))

    if size(s.energy_levels) == (0, 0)
        s.energy_levels = Matrix(undef, 0, length(energies))
        s.flashed = Matrix(undef, 0, length(energies))
    end

    s.energy_levels = [s.energy_levels; energies']
    s.flashed = [s.flashed; falses(length(energies))']
end

function simulate_step!(s::Submarine)
    (ysize, xsize) = size(s.energy_levels)

    for y = 1:ysize
        for x = 1:xsize
            increase_energy_level!(s, x, y)
        end
    end

    all_flashed_this_turn = true

    for y = 1:ysize
        for x = 1:xsize
            if s.flashed[y, x]
                s.flashed[y, x] = false
                s.energy_levels[y, x] = 0
            else
                all_flashed_this_turn = false
            end
        end
    end

    all_flashed_this_turn
end

function increase_energy_level!(s::Submarine, x, y)
    (ysize, xsize) = size(s.energy_levels)

    if x < 1 || x > xsize || y < 1 || y > ysize
        return
    end

    s.energy_levels[y, x] += 1

    if s.energy_levels[y, x] > 9
        flash!(s, x, y)
    end
end

function flash!(s::Submarine, x, y)
    if s.flashed[y, x]
        return
    end

    s.flashed[y, x] = true;
    s.total_flashes += 1

    for j in -1:1
        for i in -1:1
            increase_energy_level!(s, x + i, y + j)
        end
    end
end

function output(s::Submarine)
    i = 0
    total_flashes_after_100_steps = 0
    first_synchronized_flash = 0

    while total_flashes_after_100_steps == 0 || first_synchronized_flash == 0
        i += 1

        all_flashed = simulate_step!(s)

        if i == 100
            total_flashes_after_100_steps = s.total_flashes
        end

        if all_flashed
            first_synchronized_flash = i
        end
    end

    println("Part 1: $(total_flashes_after_100_steps)")
    println("Part 2: $(first_synchronized_flash)")
end

end

File: rust/src/day11.rs
pub struct Submarine {
    energy_levels: Vec<Vec<u32>>,
    flashed: Vec<Vec<bool>>,
    total_flashes: u32,
    width: usize,
    iwidth: i32,
    height: usize,
    iheight: i32,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            energy_levels: Vec::new(),
            flashed: Vec::new(),
            total_flashes: 0,
            width: 0,
            iwidth: 0,
            height: 0,
            iheight: 0,
        }
    }

    pub fn on_input(&mut self, result: &str) {
        let energies: Vec<u32> = result.chars().map(|c| c.to_digit(10).unwrap()).collect();

        if self.width == 0 {
            self.width = energies.len();
            self.iwidth = self.width as i32;
        }

        self.energy_levels.push(energies);
        self.flashed.push(vec![false; self.width]);

        self.height += 1;
        self.iheight += 1;
    }

    fn simulate_step(&mut self) {
        for x in 0..self.iwidth {
            for y in 0..self.iheight {
                self.increase_energy_level(x, y);
            }
        }

        // Resetting any place that flashed.
        for x in 0..self.width {
            for y in 0..self.height {
                if self.flashed[y][x] {
                    self.flashed[y][x] = false;
                    self.energy_levels[y][x] = 0;
                }
            }
        }
    }

    fn increase_energy_level(&mut self, x: i32, y: i32) {
        if x < 0 || x >= self.iwidth || y < 0 || y >= self.iheight {
            return;
        }

        self.energy_levels[y as usize][x as usize] += 1;

        if self.energy_levels[y as usize][x as usize] > 9 {
            self.flash(x, y);
        }
    }

    fn flash(&mut self, x: i32, y: i32) {
        if self.flashed[y as usize][x as usize] {
            return;
        }

        self.flashed[y as usize][x as usize] = true;
        self.total_flashes += 1;

        for i in -1..=1 {
            for j in -1..=1 {
                self.increase_energy_level(x + i, y + j);
            }
        }
    }

    pub fn output(&mut self) {
        for _ in 0..100 {
            self.simulate_step();
        }

        println!("Part 1: {}", self.total_flashes);
    }
}